### PR TITLE
add explicit fedora and centos based builds

### PR DESCRIPTION
**Pull request by obnoxxx**
This generalizes the build flavors to support multiple operating systems and adds explicit choices for fedora and centos stream9

### CODE REVIEW

File: Makefile
Comment (phlogistonjohn) [inline Makefile:45]: Extra slash after quay.io

File: Makefile
Comment (phlogistonjohn) [inline Makefile:50]: I'm not too keen on changing the image names. I'll elaborate more in the summary.

File: Makefile.opensuse
Comment (phlogistonjohn) [inline Makefile.opensuse:19]: These lines change the name of the repo at opensuse registry. I don't think this is something we can do without cooperation from dmulder nor is it something I want to do.

File: images/server/install-packages.sh
Comment (phlogistonjohn) [inline images/server/install-packages.sh:9]: Just like in #98 these script file changes will break the fedora build.  Either we need a 2nd script OR we need to generalize the script to support both centos and fedora. I prefer the latter (and am willing to make the changes).

**Review by phlogistonjohn — CHANGES_REQUESTED** [review]
First off, I'm generally in favor of making multiple base distro support follow a common pattern and not just become a one-off for opensuse support.

However, I find a lot the changes in this PR a bit confusing. The part I will focus on are the changes to the repo names to include base os names. This is a huge change and not one I think we should make. Currently we roughly have a 1:1 mapping of 'image' dir to image repo:
* images/ad-server -> quay.io/samba.org/samba-ad-server
* images/client ->  quay.io/samba.org/samba-client
* images/server -> quay.io/samba.org/samba-server
* images/toolbox -> quay.io/samba.org/samba-toolbox

(I'm going to call an abstraction of these dirs that produce an image type the image's "role" below)

What this change would lead to is a proliferation of repos, numbering `baseos * role`.  We currently treat nightly builds as a variation on the role; basically the image fulfills the same role but uses dev builds of samba.  I would:
a) Start simple by only **adding** centos, not fedora (leave fedora in the base for now)
b) vary the image tags not the repo names
c) leave opensuse makefile alone for now

I also have a personal preference that **if** we can share a Containerfile between centos and fedora without a lot of hassle we should do so, but I'm not as adamant about that.

File: Makefile.opensuse
Comment (obnoxxx) [reply Makefile.opensuse:19]: > These lines change the name of the repo at opensuse registry. I don't think this is something we can do without cooperation from dmulder nor is it something I want to do.

I agree, I meant to change this before pushing, but I seem to have forgotten,,, workin on fixing it now

**Comment by obnoxxx** [comment]
Updated the PR to address all comments/requests, I think.

now all containers build locally for all OS variants. The pattern is:

* `make -f Makefile.os-flavor  clean build-type`

e.g. : 

```make -f Makefile.fedora  clean build-client
     make -f Makefile.centos  clean build-server
make -f Makefile.opensuse  clean build-toolbox
make -f Makefile.fedora  clean build-nightly-server
```

etc ...

@phlogistonjohn - fyi, I have included the script fix with your authorship ...

**Comment by obnoxxx** [comment]
@phlogistonjohn - polished andupdated according to our discussion

**Comment by obnoxxx** [comment]
@phlogistonjohn  wrote:
> First off, I'm generally in favor of making multiple base distro support follow a common pattern and not just become a one-off for opensuse support.

Agreed

> However, I find a lot the changes in this PR a bit confusing. The part I will focus on are the changes to the repo names to include base os names. This is a huge change and not one I think we should make. Currently we roughly have a 1:1 mapping of 'image' dir to image repo:
> 
> * images/ad-server -> quay.io/samba.org/samba-ad-server
> * images/client ->  quay.io/samba.org/samba-client
> * images/server -> quay.io/samba.org/samba-server
> * images/toolbox -> quay.io/samba.org/samba-toolbox
> 
> (I'm going to call an abstraction of these dirs that produce an image type the image's "role" below)
> 
> What this change would lead to is a proliferation of repos, numbering `baseos * role`. We currently treat nightly builds as a variation on the role; basically the image fulfills the same role but uses dev builds of samba. I would:

a) Start simple by only **adding** centos, not fedora (leave fedora in the base for now)

I added both woth PR #98  in mind to make it possible to still explicitly choose fedora even if centos is made the default for server ... I won't necessarily insist on this though. Just expaining my motivation here ...

 b) vary the image tags not the repo names 

This is the case in the latest update of this PR.  The OS is only in the tag, not the repo name now.

However, I followed the pattern introduced by the opensuse build of appending `-container` to the role in the repo name, i.e.:
* `samba-server-container`
* `samba-ad-server-container`
*  `samba-client-container`
 etc. ...
@phlogistonjohn - what do you think? should we keep this or get rid of it? ...

c) leave opensuse makefile alone for now

I  essentially only touched it  to remove duplicate variable settings and to play with the newly introduced abstraction with `OS_NAME` etc ...
 
> I also have a personal preference that **if** we can share a Containerfile between centos and fedora without a lot of hassle we should do so, but I'm not as adamant about that.

**Comment by obnoxxx** [comment]
@phlogistonjohn  rebased on current master and reworked a bit after #109 had merged. I hope this was a bit more according your requests and our discussion...

File: Makefile
Comment (obnoxxx) [inline Makefile:46]:
```suggestion
NIGHTLY_SERVER_NAME=samba-server:$(NIGHTLY_TAG)
```

**Comment by obnoxxx** [comment]
Updated with a few fixed to variable expansion problems that let the builds fail locally for me. Last time I tried, I could now build locally on fedora, centos, and opensuse with both docker and podman. Currently testing again. ... 
@phlogistonjohn - the mentioned fixes are the ones we collaborated on and are squashed into the generalize commit.

File: Makefile
Comment (phlogistonjohn) [inline Makefile:21]: I don't think these changes are necessary.

File: Makefile
Comment (phlogistonjohn) [inline Makefile:26]: Same. I don't think these changes are needed.

File: Makefile
Comment (phlogistonjohn) [inline Makefile:40]: I strongly suggest we leave these values *unset* for the default makefile
```suggestion
# Set OS_NAME and OS_PREFIX in distro base specific makefiles as needed
OS_NAME=
OS_PREFIX=
```

File: Makefile.centos
Comment (phlogistonjohn) [inline Makefile.centos:3]: Add:
OS_PREFIX=$(OS_NAME)-

File: Makefile.opensuse
Comment (phlogistonjohn) [inline Makefile.opensuse:3]: We're going to have to do some extra work in this file if we don't want to change the global values.
I realized just today that setting TAG=opensuse-latest will *not match* what @dmulder originally added.
Instead of using the inherited TAG just use "latest" or use an alternate variables for the <X>_REPO_NAME variables.  The local names can continue to use TAG specifically because they're local.  I'll note that the opensuse build does not include the nightly variants so it simplifies things slightly.

File: images/server/install-sambacc.sh
Comment (phlogistonjohn) [inline images/server/install-sambacc.sh:37]: I'm not sure where this crept in from. Does removing this break a build?

File: Makefile.fedora
Comment (phlogistonjohn) [inline Makefile.fedora:3]: Add:
OS_PREFIX=$(OS_NAME)-

**Review by phlogistonjohn — CHANGES_REQUESTED** [review]
The PR is currently failng the CI because it changes both the local and global (fully qualified) container image names.  As I've suggested previously I would like to keep all existing (Makefile & Makefile.opensuse) global names the same. We can change local names if needed because they only occur on dev machines or CI nodes. 

We can keep the Makefile.centos and Makefile.fedora generating tags like "centos-latest" and "fedora-latest" but we need continuity with the project has been producing already and so "Makefile" needs to produce tags like "latest" and "nightly" without the distro prefixes.

Does that all make sense?

File: images/server/install-packages.sh
Comment (obnoxxx) [reply images/server/install-packages.sh]: the latter has happened (in collaboration ...)

File: Makefile.opensuse
Comment (dmulder) [reply Makefile.opensuse]: These repo names are actually set by the `org.opensuse.reference` label in Containerfile.opensuse in our build system. The only time we use these `*_REPO_NAME` labels is for local builds. So changing them wont break our builds in any way.

This brings up an interesting problem... Consider what happens when somebody does a `make -f Makefile.opensuse build-server` followed by a `make push-server` (without the opensuse Makefile). Wont this push an opensuse image to quay.io, overwriting the fedora image?

Comment (phlogistonjohn) [reply Makefile.opensuse]: > These repo names are actually set by the `org.opensuse.reference` label in Containerfile.opensuse in our build system. The only time we use these `*_REPO_NAME` labels is for local builds. So changing them wont break our builds in any way.
> 
That's helpful, thanks!

> This brings up an interesting problem... Consider what happens when somebody does a `make -f Makefile.opensuse build-server` followed by a `make push-server` (without the opensuse Makefile). Wont this push an opensuse image to quay.io, overwriting the fedora image?

I've been working with @obnoxxx on this PR and what I'm aiming for is to have the "distro variants", the `Makefile -f Makefile.<distro>` makefiles produce tags that include the distro name, like "fedora-latest", "fedora-nightly", "centos-latest", "opensuse-latest".  The default makefile will continue to produce tags "latest" and "nightly".  These tags apply to the "short name" and the "repo name".

So I think we'd be OK since running `make -f Makefile.opensuse build-server && make push-server` would build images named `samba-server:opensuse-latest` and `registry.opensuse.org/opensuse/samba-server:latest` [1] but the push will attempt to push `quay.io/samba.org/samba-server:latest` (which may or may not exist in the local cache, but shouldn't conflct with the opensuse build).

At least that's the intent :-)

[1] - Based on  suggestions @obnoxxx  and I  discussed yesterday on a 1:1 call.  Those changes are not pushed yet AFAICT

**Comment by obnoxxx** [comment]
this is now based on PR #110

**Comment by obnoxxx** [comment]
updated this PR after #110  has been merged

File: Makefile
Comment (obnoxxx) [reply Makefile:24]: I think those were committed by accident, and they have been undone in latest updates

**Comment by obnoxxx** [comment]
@phlogistonjohn  - I am a bit confused which change requests here are still active and which are stale. I think most have been addressed ...

**Comment by phlogistonjohn** [comment]
Job `Samba Container Image CI / build-nightly-server (pull_request)` failed because make build-nightly-server created an ad server (nightly) instead of an actual (file) server image. I'll go thru the PR and find out why.

File: Makefile
Comment (phlogistonjohn) [inline Makefile:61]:
```suggestion
NIGHTLY_SERVER_NAME=samba-server:$(NIGHTLY_TAG)
```

File: Makefile
Comment (phlogistonjohn) [inline Makefile:88]: Merge conflict marker.

File: Makefile
Comment (phlogistonjohn) [inline Makefile]: Unnecessary whitespace change

File: images/server/install-packages.sh
Comment (phlogistonjohn) [inline images/server/install-packages.sh]: I think it would be good to squash this entire patch with the previous patch. It would "undo" some of the unwanted changes that I previously commented on.

File: images/server/install-sambacc.sh
Comment (phlogistonjohn) [reply images/server/install-sambacc.sh:37]: The patch I helped you add during one of our 1:1s seems to have disappeared. There should be a patch that removes `--skip-broken --nobes` from this line. I expected to see it and comment that you could squash it with this patch. However, I can not find it on the PR.

**Review by phlogistonjohn — DISMISSED** [review]
I'll repeat myself when I recommended that you squash all the patches for this PR together.  I think you'll find it much easier to manage and given that the changes to add Centos and use OS specific variables are pretty interelated I don't think it loses much structure or "cleanness" in this case.

File: images/server/install-sambacc.sh
Comment (obnoxxx) [reply images/server/install-sambacc.sh]: I am also confused and can't explain where those changes have disappeared ... I removed the `--skip-broken` and `--nobest` parameters again and squashed with the main patch. what is still missing now are the changes to the containerfile for building sambacc during our container build process. I'll try to rediscover them

File: images/server/install-packages.sh
Comment (obnoxxx) [reply images/server/install-packages.sh]: @phlogistonjohn if I see it right, the previous patch is the one that adds the capability to explicitly choose  centos and fedora as base OS, right? Shouldn't I rather squash with the commit that introduces the script in the first place?

File: Makefile
Comment (phlogistonjohn) [reply Makefile]: Probably doesn't matter as the next patch cleans this up anyway (when debug-vars gets added).

**Comment by phlogistonjohn** [comment]
After going over this live another time the PR now looks to be in a shape I'd be OK merging. My main goal was to avoid changing any of the defaults from either running `make` or `make -f Makefile.opensuse` this meant limiting some of the changes so that it's not 100% generic but it's a good foundation.

@anoopcs9 @synarete PTAL when you have a moment

File: Makefile
Comment (anoopcs9) [inline Makefile:36]: Unwanted space at the start of line.

File: Makefile
Comment (anoopcs9) [inline Makefile:57]: Trailing whitespace towards the end of line.

File: Makefile
Comment (anoopcs9) [inline Makefile:59]: Extra new line.

File: Makefile
Comment (anoopcs9) [inline Makefile:46]: Extra new line.

File: images/server/install-sambacc.sh
Comment (anoopcs9) [inline images/server/install-sambacc.sh:37]: Extra space before `-y`. This change can be taken out.

File: images/server/install-packages.sh
Comment (anoopcs9) [inline images/server/install-packages.sh:18]: We'll have to move the above `OS_BASE` calculation towards the beginning of the script(before _get_custom_repo()_). Because installation of curl is only needed for Fedora as CentOS Stream 9 images comes with _curl-minimal_ pre-installed which conflicts with original _curl_ package.

**Review by anoopcs9 — CHANGES_REQUESTED** [review]

File: Makefile
Comment (obnoxxx) [reply Makefile:57]: @anoopcs9  sorry, I currently don't see that context in my local copy ...

File: images/server/install-packages.sh
Comment (phlogistonjohn) [reply images/server/install-packages.sh:18]: I don't know why @obnoxxx local build didn't hit that error. But rather than that change could we do `dnf install -y /usr/bin/curl` instead? I tried it on containers of both `registry.fedoraproject.org/fedora:37` and `quay.io/centos/centos:stream9`  (by hand) and in both cases it did the right thing - nothing on centos stream9 and installed curl on fedora.

File: Makefile
Comment (phlogistonjohn) [reply Makefile:36]:
```suggestion
	$(SERVER_DIR)/install-packages.sh \
```

File: Makefile
Comment (phlogistonjohn) [reply Makefile:57]:
```suggestion
NIGHTLY_TAG=$(OS_PREFIX)nightly
```

File: images/server/install-sambacc.sh
Comment (phlogistonjohn) [reply images/server/install-sambacc.sh:37]:
```suggestion
        dnf install -y "${rpmfiles[0]}"
```

File: Makefile
Comment (phlogistonjohn) [reply Makefile:59]: Let's not worry too much about extra blank lines. They're hard for @obnoxxx to find visually and they don't matter much in the grand scheme of things. You, synarete, or I can come in another day and clean them up (for example when we add support for different architectures)

File: images/server/install-packages.sh
Comment (anoopcs9) [reply images/server/install-packages.sh:18]: `dnf install /usr/bin/curl` also works.

**Comment by phlogistonjohn** [comment]
@mergifyio rebase

**Comment by obnoxxx** [comment]
@Mergifyio rebase

**Review by anoopcs9 — CHANGES_REQUESTED** [review]

File: images/server/install-packages.sh
Comment (obnoxxx) [reply images/server/install-packages.sh:18]: @anoopcs9  writes:
> `dnf install /usr/bin/curl` also works.

This is what I just added to the PR ... @phlogistonjohn - FYI

Comment (obnoxxx) [reply images/server/install-packages.sh:18]: > @obnoxxx Above change is still required.

I did the `/usr/bin/curl` thing @anoopcs9

File: images/server/install-packages.sh
Comment (obnoxxx) [inline images/server/install-packages.sh:9]: @anoopcs9  - here it is ... 😄

**Review by phlogistonjohn — COMMENTED** [review]
LGTM.  I'm not giving an approval because I ended up Co-Authoring a lot of this with @obnoxxx so I'd prefer that two others review. I will only approve it myself if the review period goes on too long.

**Review by anoopcs9 — APPROVED** [review]
lgtm, thanks.

**Comment by obnoxxx** [comment]
@phlogistonjohn  wrote:
> I'll repeat myself when I recommended that you squash all the patches for this PR together. I think you'll find it much easier to manage and given that the changes to add Centos and use OS specific variables are pretty interelated I don't think it loses much structure or "cleanness" in this case.

do you still request to squash everything?
I squashed as much as I could stand 😜

**Comment by phlogistonjohn** [comment]
> do you still request to squash everything? I squashed as much as I could stand stuck_out_tongue_winking_eye

It was not a command, it was just a suggestion.   I'm fine with it as it is.  See https://github.com/samba-in-kubernetes/samba-container/pull/107#pullrequestreview-1383998377

File: Makefile
Comment (synarete) [inline Makefile:95]: IMHO `print-vars` or `report-vars` if a better target name.

File: Makefile
Comment (synarete) [inline Makefile:129]: I prefer using the built-in [info](https://www.gnu.org/software/make/manual/make.html#index-info) function for such printing. It is just a matter of style -- no changes need here.

File: images/client/Containerfile.centos
Comment (synarete) [inline images/client/Containerfile.centos:10]: Following the same LABELs we use in `samba-operrator`, please add also `org.opencontainers.image.url="https://github.com/samba-in-kubernetes/samba-container`

File: images/server/Containerfile.centos
Comment (synarete) [inline images/server/Containerfile.centos:27]: Following the same LABELs we use in samba-operrator, please add also org.opencontainers.image.url="https://github.com/samba-in-kubernetes/samba-container

**Review by synarete — COMMENTED** [review]
LGTM

**Comment by obnoxxx** [comment]
> I'll repeat myself when I recommended that you squash all the patches for this PR together. I think you'll find it much easier to manage and given that the changes to add Centos and use OS specific variables are pretty interelated I don't think it loses much structure or "cleanness" in this case.

@phlogistonjohn  is this request still active? I thought I had squashed according to our discussion, and as much as I coud stand ... 😉

**Comment by phlogistonjohn** [comment]
> > I'll repeat myself when I recommended that you squash all the patches for this PR together. I think you'll find it much easier to manage and given that the changes to add Centos and use OS specific variables are pretty interelated I don't think it loses much structure or "cleanness" in this case.
> 
> @phlogistonjohn is this request still active? I thought I had squashed according to our discussion, and as much as I could stand ... wink

I alredy answered this question:
https://github.com/samba-in-kubernetes/samba-container/pull/107#issuecomment-1508921957

**Comment by obnoxxx** [comment]
@synarete  wrote:
> LGTM

does this comment relate to any review action? 😉

File: images/server/Containerfile.centos
Comment (obnoxxx) [reply images/server/Containerfile.centos:29]: @synarete - fwiw, I did this in a separate commit and only changed the centos and fedora based images for a start. opensuse can and probably should be treated separately. what do you think?

File: Makefile
Comment (obnoxxx) [reply Makefile:95]: @synarete wrote:

> IMHO `print-vars` or `report-vars` if a better target name.

good idea! are you ok with changing this later in a follow-up PR?

File: Makefile
Comment (obnoxxx) [reply Makefile:129]: @synarete  wrote:
> I prefer using the built-in [info]
(https://www.gnu.org/software/make/manual/make.html#index-info) function for such printing. It is just a matter of style -- no changes need here.

interesting idea again! I will also look at this later, separately. ...

**Comment by obnoxxx** [comment]
fixed one bug in a containerfile. now all builds that were failing in the CI are succeeding locally for me

**Comment by obnoxxx** [comment]
amended the label commit once more to fix the toolbox build

File: images/server/Containerfile.centos
Comment (synarete) [reply images/server/Containerfile.centos:29]: Fair enough

**Review by synarete — APPROVED** [review]

**Review by anoopcs9 — APPROVED** [review]

**Pull request merged**
